File: Computer-Based-Test-Website/API/requests/login_request.go
package requests

import (
    "errors"
    "strings"

    "API/models"
    "gorm.io/gorm" 
)

type LoginRequest struct {
    Identifier string `json:"identifier" binding:"required"`
    Password   string `json:"password" binding:"required"`
}

func (lr *LoginRequest) Authenticate(db *gorm.DB) (*models.User, error) {
    var user models.User
    if strings.Contains(lr.Identifier, "@") {
        if err := db.Where("email = ?", lr.Identifier).First(&user).Error; err != nil {
            return nil, errors.New("user not found")
        }
    } else {
        if err := db.Where("nis = ? OR nip = ?", lr.Identifier, lr.Identifier).First(&user).Error; err != nil {
            return nil, errors.New("user not found")
        }
    }
    if !CheckPassword(user, lr.Password) {
        return nil, errors.New("authentication failed")
    }

    return &user, nil
}
func CheckPassword(user models.User, password string) bool {
    return user.Password == password 
}

File: Computer-Based-Test-Website/computer-based-test/app/Http/Requests/Auth/LoginRequest.php
<?php

namespace App\Http\Requests\Auth;

use Illuminate\Auth\Events\Lockout;
use Illuminate\Foundation\Http\FormRequest;
use Illuminate\Support\Facades\Auth;
use Illuminate\Support\Facades\RateLimiter;
use Illuminate\Support\Str;
use Illuminate\Validation\ValidationException;
use App\Models\User; // Pastikan untuk mengimpor model User

class LoginRequest extends FormRequest
{
    public function authorize(): bool
    {
        return true; // Mengizinkan semua pengguna untuk melakukan request ini
    }

    public function rules(): array
    {
        return [
            'identifier' => ['required', 'string'], // Ganti 'email' dengan 'identifier'
            'password' => ['required', 'string'],
        ];
    }

    public function authenticate(): void
    {
        $this->ensureIsNotRateLimited();

        // Mencari user berdasarkan identifier (email, NIS, atau NIP)
        $user = User::where('email', $this->input('identifier'))->first();

        // Jika user tidak ditemukan atau password tidak cocok
        if (!$user || !Auth::attempt(['email' => $this->input('identifier'), 'password' => $this->input('password')], $this->boolean('remember'))) {
            RateLimiter::hit($this->throttleKey());

            throw ValidationException::withMessages([
                'identifier' => trans('auth.failed'), // Pesan error jika login gagal
            ]);
        }

        RateLimiter::clear($this->throttleKey());
    }

    public function ensureIsNotRateLimited(): void
    {
        if (!RateLimiter::tooManyAttempts($this->throttleKey(), 5)) {
            return;
        }

        event(new Lockout($this));

        $seconds = RateLimiter::availableIn($this->throttleKey());

        throw ValidationException::withMessages([
            'identifier' => trans('auth.throttle', [
                'seconds' => $seconds,
                'minutes' => ceil($seconds / 60),
            ]),
        ]);
    }

    public function throttleKey(): string
    {
        return Str::transliterate(Str::lower($this->string('identifier')).'|'.$this->ip());
    }
}